File: smalivm/src/main/java/org/cf/smalivm/opcode/IfOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class IfOp extends MethodStateOp {

    private static final Logger log = LoggerFactory.getLogger(IfOp.class.getSimpleName());
    private final IfType ifType;
    private final int register1;
    private final MethodLocation target;
    private boolean compareToZero;
    private int register2;

    IfOp(MethodLocation location, MethodLocation child, IfType ifType, MethodLocation target, int register1) {
        super(location, new MethodLocation[] { child, target });
        this.ifType = ifType;
        this.target = target;
        this.register1 = register1;
        compareToZero = true;
    }

    IfOp(MethodLocation location, MethodLocation child, IfType ifType, MethodLocation target, int register1,
         int register2) {
        this(location, child, ifType, target, register1);
        this.register2 = register2;
        compareToZero = false;
    }

    private static boolean isTrue(IfType ifType, int cmp) {
        boolean isTrue = false;
        switch (ifType) {
            case EQUAL:
                isTrue = cmp == 0;
                break;
            case GREATER:
                isTrue = cmp == 1;
                break;
            case GREATER_OR_EQUAL:
                isTrue = cmp >= 0;
                break;
            case LESS:
                isTrue = cmp == -1;
                break;
            case LESS_OR_EQUAL:
                isTrue = cmp <= 0;
                break;
            case NOT_EQUAL:
                isTrue = cmp != 0;
                break;
        }

        return isTrue;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        HeapItem lhsItem = mState.readRegister(register1);
        HeapItem rhsItem = compareToZero ? new HeapItem(0, "I") : mState.readRegister(register2);

        // Ambiguous predicate. Return to add both possible branches as children.
        if (lhsItem.isUnknown() || rhsItem.isUnknown()) {
            return;
        }

        Object lhs = lhsItem.getValue();
        Object rhs = rhsItem.getValue();
        int cmp;
        if (compareToZero) {
            if (lhs == null) {
                // if-*z ops are used to check for null refs
                cmp = 0;
            } else if ((lhs instanceof Number || lhs instanceof Boolean || lhs instanceof Character) &&
                       (rhs instanceof Number || rhs instanceof Boolean || rhs instanceof Character)) {
                Integer aIntValue = Utils.getIntegerValue(lhs);
                cmp = aIntValue.compareTo((Integer) rhs);
            } else {
                cmp = lhs == rhs ? 0 : 1;
            }
        } else if ((lhs instanceof Number || lhs instanceof Boolean || lhs instanceof Character) &&
                   (rhs instanceof Number || rhs instanceof Boolean || rhs instanceof Character)) {
            Integer aIntValue = Utils.getIntegerValue(lhs);
            Integer bIntValue = Utils.getIntegerValue(rhs);
            cmp = aIntValue.compareTo(bIntValue);
        } else {
            cmp = lhs == rhs ? 0 : 1;
        }

        if (log.isTraceEnabled()) {
            log.trace("IF compare: {} vs {} = {}", lhs, rhs, cmp);
        }

        int childIndex = isTrue(ifType, cmp) ? 1 : 0;
        node.setChildLocations(getChildren()[childIndex]);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(getName());
        sb.append(" r").append(register1);
        if (!compareToZero) {
            sb.append(", r").append(register2);
        }
        sb.append(", :addr_").append(target.getCodeAddress());

        return sb.toString();
    }

    enum IfType {
        EQUAL, GREATER, GREATER_OR_EQUAL, LESS, LESS_OR_EQUAL, NOT_EQUAL
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/IfOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.CommonTypes
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.OpChild
import org.cf.smalivm2.Value
import org.cf.util.Utils
import org.jf.dexlib2.builder.BuilderInstruction
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.OffsetInstruction
import org.jf.dexlib2.iface.instruction.OneRegisterInstruction
import org.jf.dexlib2.iface.instruction.formats.Instruction22t
import org.slf4j.LoggerFactory

class IfOp internal constructor(
    location: MethodLocation,
    child: MethodLocation,
    private val ifType: IfType,
    private val target: MethodLocation,
    private val register1: Int
) : Op(location, arrayOf(child, target)) {
    private var compareToZero = true
    private var register2 = 0

    internal constructor(
        location: MethodLocation, child: MethodLocation, ifType: IfType, target: MethodLocation, register1: Int,
        register2: Int
    ) : this(location, child, ifType, target, register1) {
        this.register2 = register2
        compareToZero = false
    }

    override val registersReadCount = if (compareToZero) 1 else 2
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out OpChild> {
        val lhs = node.state.readRegister(register1)
        val rhs = if (compareToZero) Value.wrap(0, CommonTypes.INTEGER) else node.state.readRegister(register2)

        // Ambiguous predicate. Return to add both possible branches as children.
        if (lhs.isUnknown || rhs.isUnknown) {
            return collectChildren()
        }
        val cmp = if (compareToZero) {
            if (lhs.value == null) {
                // if-*z ops are used to check for null refs
                0
            } else if ((lhs.value is Number || lhs.value is Boolean || lhs.value is Char) &&
                (rhs.value is Number || rhs.value is Boolean || rhs.value is Char)
            ) {
                val aIntValue = lhs.toInteger()
                aIntValue.compareTo(rhs.value as Int)
            } else {
                if (lhs.value === rhs.value) 0 else 1
            }
        } else if ((lhs.value is Number || lhs.value is Boolean || lhs.value is Char) &&
            (rhs.value is Number || rhs.value is Boolean || rhs.value is Char)
        ) {
            val aIntValue = lhs.toInteger()
            val bIntValue = rhs.toInteger()
            aIntValue.compareTo(bIntValue)
        } else {
            if (lhs.value === rhs.value) 0 else 1
        }

        log.trace("IF compare: {} vs {} = {}", lhs.value, rhs.value, cmp)
        val childIndex = if (isTrue(ifType, cmp)) 1 else 0
        return collectChildren(children[childIndex])
    }

    override fun toString(): String {
        val sb = StringBuilder(name)
        sb.append(" r").append(register1)
        if (!compareToZero) {
            sb.append(", r").append(register2)
        }
        sb.append(", :addr_").append(target.codeAddress)
        return sb.toString()
    }

    internal enum class IfType {
        EQUAL, GREATER, GREATER_OR_EQUAL, LESS, LESS_OR_EQUAL, NOT_EQUAL
    }

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(IfOp::class.java.simpleName)
        private fun isTrue(ifType: IfType?, cmp: Int): Boolean {
            var isTrue = false
            when (ifType) {
                IfType.EQUAL -> isTrue = cmp == 0
                IfType.GREATER -> isTrue = cmp == 1
                IfType.GREATER_OR_EQUAL -> isTrue = cmp >= 0
                IfType.LESS -> isTrue = cmp == -1
                IfType.LESS_OR_EQUAL -> isTrue = cmp <= 0
                IfType.NOT_EQUAL -> isTrue = cmp != 0
            }
            return isTrue
        }

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instruction = location.instruction as BuilderInstruction
            val address = instruction.location.codeAddress
            val branchOffset = (instruction as OffsetInstruction?)!!.codeOffset
            val targetAddress = address + branchOffset
            val child = Utils.getNextLocation(location, addressToLocation)
            val target = addressToLocation[targetAddress]!!
            val opName = instruction.getOpcode().name
            val ifType = getIfType(opName)
            val register1 = (instruction as OneRegisterInstruction?)!!.registerA
            return if (instruction is Instruction22t) {
                // if-* vA, vB, :label
                IfOp(location, child, ifType, target, register1, instruction.registerB)
            } else {
                // if-*z vA, :label (Instruction 21t)
                IfOp(location, child, ifType, target, register1)
            }
        }

        private fun getIfType(opName: String): IfType {
            return when {
                opName.contains("-eq") -> IfType.EQUAL
                opName.contains("-ne") -> IfType.NOT_EQUAL
                opName.contains("-lt") -> IfType.LESS
                opName.contains("-le") -> IfType.LESS_OR_EQUAL
                opName.contains("-gt") -> IfType.GREATER
                opName.contains("-ge") -> IfType.GREATER_OR_EQUAL
                else -> throw RuntimeException("Unrecognized IF opcode type: $opName")
            }
        }
    }
}
